### ISSUE TITLE
Siamese Network

### ISSUE THREAD

**Issue by jmschrei** [dmlc/mxnet#772]
Hello!

I am trying to use MXNet to build a Siamese network, where two networks share the same weights, and are optimized using a contrastive loss function, rather than softmax or MSE. I was looking for some insight into how to do this. A simplified version of my code is below.

```
def AugConvolution( data, num_filter, kernel, stride=(1,1), pad=(0,0), name=None, suffix='' ):
	"""
	Create a convolution layer with batch normalization and relu activations.
	"""

	conv = Convolution( data=data, num_filter=num_filter, kernel=kernel, stride=stride, pad=pad, name='conv_{}{}'.format( name, suffix ) )
	bn = BatchNorm( data=conv, name='bn_{}{}'.format( name, suffix ) )
	act = Activation( data=bn, act_type='relu', name='relu_{}{}'.format( name, suffix ) )
	return act

def JNET():
	x1seq = Variable( name="x1seq" )
	x1seq_conv1 = AugConvolution( x1seq, 64, (5, 1), name="x1seq", suffix="conv1" )
	x1_flatten = Flatten( x1seq_conv1 )

	x2seq = Variable( name="x2seq" )
	x2seq_conv1 = AugConvolution( x2seq, 64, (5, 1), name="x2seq", suffix="conv1" )
	x2_flatten = Flatten( x2seq_conv1 )

	concat = Concat( [x1_flatten, x2_flatten] )
	fc1 = FullyConnected( concat, num_hidden=2 )
	softmax = SoftmaxOutput( data=fc1, name="softmax" )

def main():
	"""Fit some data and make predictions."""

	x1_seq, x1_dnase, x2_seq, x2_dnase, y = extract_dataset(21)
	y = y.reshape( y.shape[0], 2 ).argmax( axis=1 )

	softmax = JNET()
	model = mx.model.FeedForward( ctx=[mx.gpu(2), mx.gpu(3)], symbol=softmax, num_epoch=10, learning_rate=0.001, momentum=0.9, wd=0.00001 )
	model.fit( ?? )

if __name__ == '__main__':
	main()
```

I have multiple questions:

(1) How do I pass in the multiple inputs to the model using model.fit? I haven't been able to find anything in either the documentation or the code showing me exactly how to do that. 
(2) What is the best way to share weights between layers, particularly BatchNorm layers and PReLU layers?
(3) Are there any plans to implement a contrastive loss function soon? I know I can write my own, but would prefer a hardcoded one.

**Comment by piiswrong** [comment]
1&2) I would recommend writing your own model and solver for non-feedforward models. It basically goes like this:
sym = build_your_model()
vars = allocate_your_variables()
exe = sym.bind(...) #if you want to share weight between x1 and x2, just set args = {'x1': X, 'x2': X...} i.e. bind them to the same variable. don't forget to also bind the gradient to the same variable.
while True:
  copy_data_to_input()
  exe.forward(isTrain=True)
  exe.backward()
  for weight, grad in zip(weight_arrays, grad_arrays)
    optimizer.update(index, weight,grad)
  copy_output_to_buffer_and_eval()

you can check examples/autoencoder for reference. Also read the individual API doc.

**Comment by piiswrong** [comment]
On a second thought, this is actually a feed forward model. So if you want to use the standard API, try NDArrayIter. It takes a bunch of numpy arrays and create a dataiterator

**Comment by jmschrei** [comment]
Right. I am aware of NDArrayIter, but I am not sure how it assigns the arrays to the variables. Can you give a brief example?

**Comment by piiswrong** [comment]
by name. You allocated mx.sym.Variable('xxx') for your networks input symbol. That name must match your initialization dict for FeedForward model.

**Comment by jmschrei** [comment]
Are you saying that NDArrayIter allows X to be a dictionary?

**Comment by piiswrong** [comment]
Yes. The document seems to be outdated

**Comment by jonbakerfish** [comment]
For BatchNorm layer, do we need to share their auxiliary states as well?

**Comment by xzqjack** [comment]
Hi @piiswrong , i still have some doubts about sharing weights. Bellow is my stupide code:
```
    data1 = mx.sym.Variable("data1")
    data2 = mx.sym.Variable("data2")
    conv_weight = mx.sym.Variable("conv_weight")
    conv_bias = mx.sym.Variable("conv_bias")    
    conv_1 = mx.symbol.Convolution(data=data1, weight=conv_weight, bias=conv_bias, kernel=(3, 3), pad=(1, 1), num_filter=64)
    conv_2= mx.symbol.Convolution(data=data1, weight=conv_weight, bias=conv_bias, kernel=(3, 3), pad=(1, 1), num_filter=64)
```
I have to define two similar network with same weight, which is inefficient,  especially when the network is deep. Is there any simple way to make two same network share same weights just like
```
define network():
      ..........
net_1 = network()
net_2 = network()
share_weight(net_1,net_2)
net = mx.sym.Concat(net_1, net_2,)
```
Thanks.

**Comment by pengwangucla** [comment]
@xzqjack I have exact same problem for sharing weight. 

Say if we want to have a sharing weight resnet 101, it will be much more ugly having parameter arguments passing to every layer of blocks. 

Is is possible just passing parameter name to each convolution layer for sharing weights, just the same as every other platforms like tensorflow, caffe and paddle ? 

Thanks
